src/Avalonia.Visuals/Rendering/SceneGraph/VisualNode.cs

```diff
@@ -172,6 +172,37 @@ public void ReplaceDrawOperation(int index, IRef<IDrawOperation> operation)
             old.Dispose();
         }
 
+        /// <summary>
+        /// Sorts the <see cref="Children"/> collection according to the order of the visual's
+        /// children and their z-index.
+        /// </summary>
+        /// <param name="scene">The scene that the node is a part of.</param>
+        public void SortChildren(Scene scene)
+        {
+            var keys = new List<long>();
+
+            for (var i = 0; i < Visual.VisualChildren.Count; ++i)
+            {
+                var child = Visual.VisualChildren[i];
+                var zIndex = child.ZIndex;
+                keys.Add(((long)zIndex << 32) + i);
+            }
+
+            keys.Sort();
```

Comment:
Actually, not OK. List.Sort is not stable

Reply:
Yep, somewhere in the time frame of merging the scene graph, GetSortedVisualList (which was using List.Sort) got replaced by a [VisualExtension.SortByZIndex](https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Visuals/VisualTree/VisualExtensions.cs#L239) which does a stable sort.

Reply:
We're sorting on the index (`i`) as well as the zindex though, which means the sort doesn't need to be stable. In fact this code was taken from WPF which does something similar _because_ `List.Sort` isn't stable: https://github.com/dotnet/wpf/blob/master/src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/Controls/panel.cs#L1030-L1035

Reply:
Ah, OK